### PR TITLE
[4] Introducing empty state to modules

### PR DESCRIPTION
**Pull request by PhilETaylor**
### Summary of Changes

Introducing empty state design for modules. A lot less complex than the component versions. 

This first PoC on the 3 modules shown by default on the home page of the Joomla 4 admin after first login. Those are 

```
mod_latest
mod_popular
mod_privacy_dashboard
```

The sentence displayed can reuse existing empty state title from the component, or a new one can be overloaded for modules in the language strings 

Can handle more than one module type per component with the use of `textSuffix` 

### Testing Instructions

Load Joomla 4 admin console.

### Actual result BEFORE applying this Pull Request

Screenshot 2021-05-01 at 20 09 40

### Expected result AFTER applying this Pull Request

Screenshot 2021-05-01 at 20 09 22

### Documentation Changes Required

none

### CODE REVIEW

**Comment by brianteeman** [comment]
The strings are not correct.
No current privacy requests
There are no articles

**Comment by PhilETaylor** [comment]
"No Information Requests have been submitted yet." is the current language value from the component. So if you dont like it here, you are going to hate it there too. 

Screenshot 2021-05-01 at 21 11 58

And the two articles strings are NOT trying to say that there are no articles in the database, they are trying to say that the purpose of the module is a empty state, and as the data for these modules is about Popular and Recently added articles, the language has to reflect that underlying dataset - there are no popular articles and there are no recent articles. You could well have 1000 articles, but without 1 visitors the popular module could still be empty of data. 

While that may actually be true - Silly Joomla goes and puts articles with Zero hits in the module.. /facepalm

Screenshot 2021-05-01 at 21 14 35

**Comment by PhilETaylor** [comment]
I have removed the module override language strings for the empty state to allow them to inherit from the components empty state language tites.

Screenshot 2021-05-01 at 21 17 06

File: administrator/modules/mod_privacy_dashboard/mod_privacy_dashboard.php [hunk 1]
```diff
@@ -9,7 +9,9 @@
 
 defined('_JEXEC') or die;
 
+use Joomla\CMS\Factory;
```
Comment (Quy) [inline administrator/modules/mod_privacy_dashboard/mod_privacy_dashboard.php]:
```suggestion
```

File: layouts/joomla/content/emptystate_module.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * @package     Joomla.Site
+ * @subpackage  Layout
+ *
+ * @copyright   (C) 2021 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Factory;
```
Comment (Quy) [inline layouts/joomla/content/emptystate_module.php]:
```suggestion
```

File: administrator/modules/mod_latest/mod_latest.php [hunk 3]
```diff
@@ -20,4 +21,18 @@
 	$module->title = LatestHelper::getTitle($params);
 }
 
-require ModuleHelper::getLayoutPath('mod_latest', $params->get('layout', 'default'));
+if (!count($list))
```
Comment (Quy) [inline administrator/modules/mod_latest/mod_latest.php]: Me just being nit picky that swapping the cases to remove `!`.

Comment (PhilETaylor) [reply administrator/modules/mod_latest/mod_latest.php]: Done x 3

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on fb12689f1303391dbe0f0611d3361dc944a0532c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33472">issues.joomla.org/tracker/joomla-cms/33472</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on fb12689f1303391dbe0f0611d3361dc944a0532c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33472">issues.joomla.org/tracker/joomla-cms/33472</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33472">issues.joomla.org/tracker/joomla-cms/33472</a>.</sub>

File: layouts/joomla/content/emptystate_module.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+$textPrefix = $displayData['textPrefix'];
+$icon = $displayData['icon'] ?? 'icon-copy article';
+$componentLangString = $textPrefix . '_EMPTYSTATE_TITLE';
+$moduleLangString = $textPrefix . '_EMPTYSTATE_MODULE_TITLE' . (array_key_exists('textSuffix', $displayData) ? $displayData['textSuffix'] : '');
+$moduleTitle = Text::_($moduleLangString);
+
+$title = $displayData['title']
+		?? ($moduleTitle != $moduleLangString
+				? Text::_($moduleTitle)
+				: Text::_($componentLangString));
```
Comment (Quy) [inline layouts/joomla/content/emptystate_module.php]:
```suggestion
	?? ($moduleTitle != $moduleLangString
		? Text::_($moduleTitle)
		: Text::_($componentLangString));
```

**Comment by Quy** [comment]
Thank you!

**Pull request merged**